iceoryx_posh/source/popo/building_blocks/condition_listener.cpp

```diff
@@ -64,43 +64,37 @@ bool ConditionListener::wasNotified() const noexcept
     return *result != 0;
 }
 
-void ConditionListener::wait() noexcept
+ConditionListener::NotificationVector_t ConditionListener::wait() noexcept
 {
-    if (m_toBeDestroyed.load(std::memory_order_relaxed))
-    {
-        return;
-    }
-
-    getMembers()->m_semaphore.wait().or_else([](auto) {
-        errorHandler(Error::kPOPO__CONDITION_LISTENER_SEMAPHORE_CORRUPTED_IN_WAIT, nullptr, ErrorLevel::FATAL);
+    return waitImpl([this]() -> bool {
+        if (this->getMembers()->m_semaphore.wait().has_error())
+        {
+            errorHandler(Error::kPOPO__CONDITION_LISTENER_SEMAPHORE_CORRUPTED_IN_WAIT, nullptr, ErrorLevel::FATAL);
+            return false;
+        }
+        return true;
     });
 }
 
-bool ConditionListener::timedWait(units::Duration timeToWait) noexcept
+ConditionListener::NotificationVector_t ConditionListener::timedWait(const units::Duration& timeToWait) noexcept
 {
-    if (m_toBeDestroyed.load(std::memory_order_relaxed))
-    {
-        return false;
-    }
-
-    auto continueOnInterrupt{false};
-    auto result = getMembers()->m_semaphore.timedWait(timeToWait, continueOnInterrupt);
-
-    if (result.has_error())
-    {
-        errorHandler(Error::kPOPO__CONDITION_LISTENER_SEMAPHORE_CORRUPTED_IN_TIMED_WAIT, nullptr, ErrorLevel::FATAL);
+    return waitImpl([this, timeToWait]() -> bool {
+        if (this->getMembers()->m_semaphore.timedWait(timeToWait, true).has_error())
+        {
+            errorHandler(
+                Error::kPOPO__CONDITION_LISTENER_SEMAPHORE_CORRUPTED_IN_TIMED_WAIT, nullptr, ErrorLevel::FATAL);
+        }
         return false;
```

Comment:
I am sure it is correct.

- `ConditionNotifier::notify()` set first the `m_activeNotification[] = true` and then `posts` the semaphore
- after the `waitCall` returns in `waitImpl` another round is being done where the `activeNotifications` are collected again and returned

But it can happend that:
1. notify is triggered
2. wait awakes int waitImpl
3. notify is triggered again
4. waitImpl collects all activeNotifications and the newly triggered one

When you call wait again you receive an empty activeNotificationsVector when it is being called with timedWait. This spurious wake up is alright and can happen. If you see a quick fix you can post it here but @budrus agreed on this behavior.

Reply:
Oh and `timedWait` in the semaphore either returns a `SemaphoreWaitState` when the calls was successful. The WaitState can contain either the value `TIMEOUT` or `NO_TIMEOUT`.
If it was unsuccessful the expected contains a semaphore error.

Reply:
oh, I overlooked the reset at the beginning